### PR TITLE
Support new axes in da.atop

### PR DESCRIPTION
**Pull request by shoyer**
Yes, `da.map_blocks` already supports adding new axes. But `map_blocks` doesn't support combining arrays with different dimensions, which makes it less useful for xarray.

This will be used to enhance an "autodask" mode for applying ufunc like functions in xarray, which I think will be quite useful:
https://github.com/pydata/xarray/pull/964/files#diff-293280a1ed0154c9fb35b842f69e7e33R290

If the user supplies coordinate labels for new dimensions we also know their size, which should be sufficient information for us to call `da.atop` with `out_chunks`, assuming new dimensions consist of a single block.

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -1728,6 +1740,10 @@ def atop(func, out_ind, *args, **kwargs):
 
     >>> z = atop(sequence_dot, '', x, 'i', y, 'i')  # doctest: +SKIP
 
+    Insert a new axes in the result:
+
+    >>> z = atop(lambda x: x[..., None], 'ij', x, 'i', out_chunks={'i': 1})  # doctest: +SKIP
```
Comment (mrocklin) [inline dask/array/core.py]: should this be `{'j': 1}` ?

**Comment by mrocklin** [comment]
For me I think the question is "is adding new dimensions in this way sufficiently tied into what `atop` does that it couldn't be handled by another general operation?  There is a tension between making `atop` do everything, which is convenient downstream but makes it harder to reason about and maintain upstream, or having a few different operations.  If the functionality we want to add requires other inputs then I'm tempted to see if there is a way to separate the new functionality into a different function.  Thoughts?

**Comment by shoyer** [comment]
Sure, I'm open to other approaches here.

For xarray, we could achieve the desired functionality by simply calling `atop` and fixing up the dask keys/Array chunks after the fact, but that makes me nervous because the temporary Array is invalid. We could use `top` directly, but then would need to duplicate most of the logic from `atop`.

What I'd really like is a version of `elemwise` that supports core dimensions for input/output ala NumPy's generalized ufuncs (http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html). I think this could be useful for more than just xarray, which is why I want to push as much of this functionality as possible into dask.

**Comment by mrocklin** [comment]
> For xarray, we could achieve the desired functionality by simply calling atop and fixing up the dask keys/Array chunks after the fact, but that makes me nervous because the temporary Array is invalid. We could use top directly, but then would need to duplicate most of the logic from atop.

Yeah, short term I think I would solve this by using atop as you suggest, having chunks with more dimensions than the array itself, and then calling some `expand_dimensions` function that added extra dimensions into the array, and added dimensions into the chunks at runtime if necessary.  

However, I haven't thought about this much yet.  it could be that what's in this PR (or some slight modification of this PR) is cleaner.  I honestly haven't spend sufficient time to rule this out or in.  My apologies for the low bandwidth here.  I'm generally fairly conservative with expanding the scope of `atop`, but perhaps that's unwise.

> What I'd really like is a version of elemwise that supports core dimensions for input/output ala NumPy's generalized ufuncs (http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html). I think this could be useful for more than just xarray, which is why I want to push as much of this functionality as possible into dask.

It seems like one could do a bit with `atop` + `expand_dimensions`  or some such.  Full gufunc support is expensive to build.

**Comment by shoyer** [comment]
> Yeah, short term I think I would solve this by using atop as you suggest, having chunks with more dimensions than the array itself, and then calling some expand_dimensions function that added extra dimensions into the array, and added dimensions into the chunks at runtime if necessary.

Sounds good, this approach is pretty reasonable.

**Comment by shoyer** [comment]
In case anyone does decide to work on gufunc support in dask at some future time, I wrote a simple Python parser for NumPy's gufunc signatures for xarray: https://github.com/pydata/xarray/pull/964/files#diff-293280a1ed0154c9fb35b842f69e7e33R92

NumPy does not expose it's own code for doing this parsing, which is written in C.

**Pull request closed without merging**